args <- commandArgs(trailingOnly = TRUE)

PDIR <- args[1]
countDIR <- args[2]
tablePATH <- args[3]
JOBS <- args[4]

setwd(PDIR)

suppressMessages(suppressWarnings(library (GenomicFeatures)))
# suppressMessages(suppressWarnings(library (tximport)))
suppressMessages(suppressWarnings(library (tidyverse)))
suppressMessages(suppressWarnings(library (DESeq2)))
suppressMessages(suppressWarnings(library (BiocParallel)))
suppressMessages(suppressWarnings(library(RColorBrewer)))
suppressMessages(suppressWarnings(library(pheatmap)))
suppressMessages(suppressWarnings(library (ggplot2)))
suppressMessages(suppressWarnings(library (ggrepel)))
# suppressMessages(suppressWarnings(library (patchwork)))
suppressMessages(suppressWarnings(library(plotly)))
suppressMessages(suppressWarnings(library(htmlwidgets)))

register(MulticoreParam(JOBS))
message ('R libraries loaded!')

# make new directory 
mkdir <- function(PATH){
    ifelse(
        !dir.exists(file.path('.', PATH)), 
        dir.create(file.path('.', PATH), showWarnings = FALSE), 
        paste(PATH,'already exists!')
    )
}

ggplot_Save <- function (p, name_it){
    ggsave(paste(name_it,'png',sep='.'), plot = p, device = 'png', dpi = 300)
    ggsave(paste(name_it,'pdf',sep='.'), plot = p, device = 'pdf', dpi = 300)
}

###################################################################################################
message ('@@@@@@@@@@@@@@@@@@@@@@@@@@@@@')
message ('Part 1. Read genome annotations')
message ('@@@@@@@@@@@@@@@@@@@@@@@@@@@@@')

GTF = '/rumi/shams/abe/genomes/hg38/gencode.v34/gencode.v34.annotation.gtf'

gtf <- rtracklayer::import(GTF)
gene2name <- gtf[gtf$type == "gene"] %>% data.frame %>% column_to_rownames('gene_id') %>% dplyr::select('gene_name')

message ('-> GTF loaded!')

###################################################################################################
message ('@@@@@@@@@@@@@@@@@@@@@@@@@@@@@')
message ('Part 2: Read  count data into a DESeq object')
message ('@@@@@@@@@@@@@@@@@@@@@@@@@@@@@')

# make `deseq` directory to save resluts  
mkdir ('deseq')

# # `sampleTable` for htseq-count: a data.frame with three or more columns. Each row describes one sample. The first column is the sample name, the second column the file name of the count file generated by htseq-count, and the remaining columns are sample metadata which will be stored in colData - [link](https://www.rdocumentation.org/packages/DESeq2/versions/1.12.3/topics/DESeqDataSet-class)

# # read meta table 
# sampleTable <- read.csv( tablePATH, sep="\t" )
# # read count files 
# sampleFiles <- list.files(pattern=".txt", path=countDIR)
# # htseq-count friendly sampleTable
# sampleTable = sampleTable %>% 
#     add_column (sampleFiles, .after=1) %>% 
#     add_column(
#         group = factor(paste(sampleTable$cells,sampleTable$sorted,sampleTable$treatment, sep='.')), 
#         .after=2) # should modify

# make DESeq Obj. from htseq-counts 
ddsHTSeq <- DESeqDataSetFromHTSeqCount(
    sampleTable = sampleTable,
    directory = countDIR,
    design= ~ group # or something else
)

# run DESeq function 
dds0 <- DESeq(ddsHTSeq, parallel=TRUE)

# save merged raw counts into a file 
counts.raw <- counts(dds0) %>% data.frame %>% 
    rownames_to_column('gene_id') %>% 
    add_column(gene_name=gene2name[rownames(ddsHTSeq),], .after='gene_id')

write.table(
    counts.raw,'deseq/counts.raw.txt', 
    sep="\t", quote=FALSE, row.names = FALSE
)

# save merged normalized counts into a file
counts.normalized <- counts(dds0, normalized=TRUE) %>% data.frame %>% 
    rownames_to_column('gene_id') %>% 
    add_column(gene_name=gene2name[rownames(ddsHTSeq),], .after='gene_id')

write.table(
    counts.normalized,'deseq/counts.normalized.txt', 
    sep="\t", quote=FALSE, row.names = FALSE
)
message ('-> DESeq object loaded! Size:', dim(dds0)[1], 'x', dim(dds0)[2])


###################################################################################################
message ('@@@@@@@@@@@@@@@@@@@@@@@@@@@@@')
message ('Part 3. QC plots')
message ('@@@@@@@@@@@@@@@@@@@@@@@@@@@@@')
message ('Part 3.1. 3D PCA plot')


widget_file_size <- function(p) {
  d <- tempdir()
  withr::with_dir(d, htmlwidgets::saveWidget(p, "index.html"))
  f <- file.path(d, "index.html")
  mb <- round(file.info(f)$size / 1e6, 3)
  message("File is: ", mb," MB")
}

plotPCA3D <- function (object, color = 'cells', intgroup = 'group', ntop = 5000, returnData = TRUE){
  rv <- rowVars(assay(object))
  select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, length(rv)))]
  pca <- prcomp(t(assay(object)[select, ]))
  percentVar <- pca$sdev^2/sum(pca$sdev^2)
  if (!all(intgroup %in% names(colData(object)))) {
    stop("the argument 'intgroup' should specify columns of colData(dds)")
  }
  intgroup.df <- as.data.frame(colData(object)[, intgroup, drop = FALSE])
  group <- if (length(intgroup) > 1) {
    factor(apply(intgroup.df, 1, paste, collapse = " : "))
  }
  else {
    colData(object)[[intgroup]]
  }
  d <- data.frame(PC1 = pca$x[, 1],
                  PC2 = pca$x[, 2],
                  PC3 = pca$x[, 3],
                  group = group,
                  intgroup.df,
                  name = colnames(object))
  if (returnData) {
    attr(d, "percentVar") <- percentVar[1:3]
    return(d)
  }
}

vsd <- varianceStabilizingTransformation(ddsHTSeq)

message("Generating plotly 3-D PCA plot!")
d = plotPCA3D(vsd)

# Define the number of colors you want
nb.cols <- length(d$group)
mycolors <- colorRampPalette(brewer.pal(8, "Set2"))(nb.cols)

p = plotly::plot_ly(
  data = d,
  x = ~PC1, y = ~PC2, z = ~PC3,
  color = d$group, 
  colors = mycolors,
  mode = "markers", type = "scatter3d"
)

percentVar <- round(100 * attr(d, "percentVar"))
p <- p %>% layout(
    title = "Principal Component Analysis",
    scene = list(
        xaxis = list(title = paste0("PC1: ",percentVar[1],"% variance")),
        yaxis = list(title = paste0("PC2: ",percentVar[2],"% variance")),
        zaxis = list(title = paste0("PC3: ",percentVar[3],"% variance"))
    )
)

setwd('deseq')
widget_file_size(p)

# make `pca3d` directory to save resluts  
mkdir ('pca3d')

saveWidget(p, "pca3d.html", selfcontained = F, libdir = "pca3d")    
setwd('../')
    
###################################################################################################
message ('Part 3.2. Hierarchical clustering heatmap of most variant genes')


mostVar <- function(data, n_hits, i_want_most_var = TRUE) { 
    # https://rdrr.io/github/abc-igmm/transcripTools/man/mostVar.html
    # evaluate variance 
    data.var <- apply(data, 1, stats::var)
    # order by variance and return 
    data <- data[order(data.var, decreasing = i_want_most_var)[1:n_hits],] 
    
    return (data)
}

keep <- rowSums(counts(ddsHTSeq) > 10) == ncol(ddsHTSeq)

zscore = scale(t(counts.normalized[keep,3:ncol(counts.normalized)]))    
data = mostVar(data=t(zscore), n_hits=1000, TRUE)

ann = data.frame(
    cell=sampleTable$cells, 
    treatment=sampleTable$treatment, 
    sorted=sampleTable$sorted, 
    row.names=sampleTable$sample_ID)


rownames(ann) <- colnames(data) 

colors <- seq(-1,1,by=0.01)
my_palette <- c("dodgerblue",colorRampPalette(colors = c("dodgerblue", "ghostwhite", "firebrick1"))
                                                   (n = length(colors)-3), "firebrick1")

h1 <- pheatmap(
    data, 
    labels_row = gene2name[data %>% rownames,],
    color = my_palette, 
    breaks = colors, 
    angle_col= 45,
    cluster_col=FALSE,
    cellwidth = 12,
    show_rownames=FALSE,
    annotation_col = ann,
    show_colnames=FALSE,
    annotation_legend=TRUE,
#     margin = c(5,5)
)

ggplot_Save(h1,'deseq/mostVar_Heatmap')

# message ('@@@@@@@@@@@@@@@@@@@@@@@@@@@@@')
# message ('Part 4: Run the DESeq tests ')
# message ('@@@@@@@@@@@@@@@@@@@@@@@@@@@@@')

# plot_Volcano <- function(res, lfc.cutoff  = 1,pval.cutoff = 0.05, title=''){
#     res$sig <- as.factor(res$pvalue < pval.cutoff & abs(res$log2FoldChange) > lfc.cutoff)
#     relevel(res$sig, ref=TRUE)
#     ## Volcano plot
#     vol = res %>% ggplot(
#         aes(x=log2FoldChange, y=-log10(pvalue), colour=sig, fill=sig)) +
#         geom_point(aes(color = sig),alpha = 1/10) +
#     #         xlim(c(-20,20)) +
#     #         ylim(c(0,11)) +
#             geom_hline(yintercept=-log10(pval.cutoff), linetype="dashed", alpha = 4/10) +
#             geom_vline(xintercept=lfc.cutoff, linetype="dashed", alpha = 4/10) +
#             geom_vline(xintercept=(-1)*lfc.cutoff, linetype="dashed", alpha = 4/10) +
#             scale_color_manual(values = c("grey", "red")) +
#             theme_bw() + 
#             theme(legend.position="none") +
#             ggtitle (title) + 
#             geom_text_repel(
#                 data = subset(res[order(res$pvalue),], sig == TRUE)[1:10,],
#                 aes(label = name),
#                 size = 3,
#                 box.padding = unit(0.35, "lines"),
#                 point.padding = unit(0.3, "lines")
#             )
#      return (vol)
# }


# write_Result <- function(res, name_it, col=FALSE, row=FALSE){
#     write.table(res,name_it, sep="\t", quote=FALSE, col.names=col, row.names=row)
# }

# get_Result <- function(ddsin, condition, comp1, comp2){
    
#     model = paste(condition,comp1,'vs',comp2, sep='_')

#     colData(ddsin)[,condition] <- relevel(colData(ddsin)[,condition], ref=comp2)
    
#     ddsout <- DESeq(ddsin, parallel=TRUE)
    
#     res <- results(ddsout, name=model, parallel=TRUE)  %>%
#         data.frame %>% 
#         add_column(name=gene2name[rownames(dds),]) %>%
#         add_column(ensembl=rownames(dds) %>% substr(0, 15))

#     return (res)
# }


# mkdir ('ipage')
# mkdir ('deseq/results')

# for (i in c(1:dim(comp)[1]) ){
#     res = NA
#     res = get_Result(ddsHTSeq, 'group', comp$V1[i], comp$V2[i])

#     name = paste(comp$V1[i], 'vs', comp$V2[i], sep='_')

#     vol = plot_Volcano(res, title=name)
    
#     ggplot_Save(vol, paste('deseq/results/',name,'.Volcano-plot',sep='') )
    
#     write_Result(res[,c(7,1:6)], paste('deseq/results/',name,'.DESeq-result.txt', sep=''), row=TRUE)
    
#     # Prepare results for running iPAGE:
#     res$log2FoldChange[is.na(res$log2FoldChange)] <- 0
#     write_Result(res [,c('ensembl','log2FoldChange')] %>% remove_rownames, paste('ipage/',name,'.txt',sep=''))
# }
